sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataRetriever.java

```diff
@@ -0,0 +1,51 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+package com.azure.android.core.http.responsepaging;
+
+import com.azure.android.core.http.Callback;
+import com.azure.android.core.util.paging.Page;
+
+import java.util.Objects;
+
+/**
+ * Represents an asynchronous paged data provider with fixed page size
+ * @param <T> type of the items contained in the page
+ * @param <P> page that contains the items
+ */
+public abstract class AsyncPagedDataRetriever<T, P extends Page<T>> {
+    private final int pageSize;
+
+    /**
+     * Constructor requires the fixed page size
+     * @param pageSize size of the pages
+     */
+    public AsyncPagedDataRetriever(int pageSize){
+        Objects.requireNonNull(pageSize);
+        if (pageSize <= 0)
+            throw new IllegalArgumentException("pageSize must be a positive integer");
+        this.pageSize = pageSize;
+    }
+
+    /**
+     * Gets the size of the page
+     * @return size of the page
+     */
+    public int getPageSize(){
+        return this.pageSize;
+    }
+
+    /**
+     * Retrieve the first page in the collection
+     * @param callback callback function to handle the page data along with the response for retrieving it
+     */
+    public abstract void getFirstPage(Callback<P> callback);
+
+    /**
+     * Retrieve a page with its id
+     * @param pageId id of the page
+     * @param callback callback function to handle the page data along with the response for retrieving it
+     */
+    public abstract void getPage(String pageId, Callback<P> callback);
```

Comment:
How is the pageId supposed to be communicated to the developer? Is it an opaque identifier (e.g. GUID)? I would expect `getPage(int pageNumber, Callback callback)` or something similar, otherwise the developer has to do book-keeping.

Alternatively, why not keep the current page number in the instance's local state, and have a `getNextPage(Callback callback)` method (and a simple getter, `getCurrentPageNumber()`or some such) instead?

Reply:
I guess this comment isn't specific to this file, but just to the general design.

Reply:
For the need of autorest.android, the String id is meant to be opaque. I even had a page iterator function in the collection trying to hide the pageId altogether but it turned out to be overly complicated (in thread synchronization). So we agreed to remove the iterator implementation. 

We looked at using this interface to implement androidx.arch.DataSource in my initial Draft PR. Having an opaque page id is more convenient than int id for that particular purpose. That's the reason we settled with an opaque string id.

c.f https://developer.android.com/reference/android/arch/paging/PageKeyedDataSource#loadBefore(android.arch.paging.PageKeyedDataSource.LoadParams%3CKey%3E,%20android.arch.paging.PageKeyedDataSource.LoadCallback%3CKey,%20Value%3E)

Reply:
Thinking a bit more about it: 

The paged collection = retrieve pages from service via network + caching of the pages.

Wondering, whether it makes sense to have the following signature:

```java
// group_1: calls go over network.
void getFirstPage(Callback<Page<T>> callback);
void getNextPage(Callback<Page<T>> callback);
```

```java
// group_2: an immutable map of cached pages.
Map<String, Page<T>> getCachedPages();
// an immutable list of items in the cached pages.
List<T> getCachedItems();
```

Only the APIs that belongs to the group_1 provides access to the Response object.

Reply:
Leaving as-is for this preview.